client/store/product.js

```diff
@@ -2,29 +2,103 @@ import axios from 'axios';
 
 //Action Types
 const SET_PRODUCTS = 'SET_PRODUCTS';
+const NEW_PRODUCT = 'NEW_PRODUCT';
+const UPDATE_PRODUCT = 'UPDATE_PRODUCT';
+const REMOVE_PRODUCT = 'REMOVE_PRODUCT';
+const FETCH_REQUEST = 'FETCH_REQUEST';
+const FETCH_FAILURE = 'FETCH_FAILURE';
 
 //Action Creator
 export const setProducts = products => ({
   type: SET_PRODUCTS,
   products,
 });
 
+export const newProduct = product => ({
+  type: NEW_PRODUCT,
+  product,
+});
+
+export const removeProduct = product => ({
+  type: REMOVE_PRODUCT,
+  product,
+});
+
+export const updateProduct = product => ({
+  type: UPDATE_PRODUCT,
+  product,
+});
+
+const fetchRequest = () => ({ type: FETCH_REQUEST });
+
+const fetchFailure = error => ({
+  type: FETCH_FAILURE,
+  error: 'Failed to fetch products',
+  payload: error,
+});
+
 //Thunk Creator
 export const fetchProducts = () => dispatch => {
+  dispatch(fetchRequest());
   return axios
     .get('./api/products')
     .then(res => res.data)
     .then(products => dispatch(setProducts(products)))
     .catch(error => {
       console.error.bind(console);
+      dispatch(fetchFailure(error));
     });
 };
 
+export const addProduct = product => dispatch => {
+  return axios
+    .post('/api/products', product)
+    .then(res => dispatch(newProduct(res.data)))
+    .catch(console.error.bind(console));
+};
+
+export const deleteProduct = product => dispatch => {
+  return axios
+    .delete(`api/products/${product.id}`)
+    .then(() => dispatch(removeProdcut(product)))
+    .catch(console.error.bind(console));
+};
+
+export const sendProductUpdate = product => dispatch => {
+  return axios
+    .put(`./api/products/${product.id}`, product)
+    .then(() => dispatch(updateProduct(product)))
+    .catch(console.error.bind(console));
+};
+
+const initialState = { products: [], isFetching: false };
+
 //Reducer
-export const productsReducer = (state = [], action) => {
+export default (state = initialState, action) => {
   switch (action.type) {
+    case FETCH_REQUEST:
+      return { ...state, isFetching: true };
     case SET_PRODUCTS:
-      return action.products;
+      return { ...state, products: action.products, isFetching: false };
+    case NEW_PRODUCT:
+      return { ...state, products: [...state.products, action.product] };
+    case REMOVE_PRODUCT:
+      return {
+        ...state,
+        products: state.products
+          .slice()
+          .filter(product => product.id !== action.product.id),
+      };
+    case UPDATE_PRODUCT:
+      return {
+        ...state,
+        products: [
+          ...state.products
+            .slice()
+            .filter(product => product.id !== product.id),
+          action.product,
+        ],
+      };
     default:
```

Comment:
Should we have another action.type for Fetch Failure as well? I think we do, but I am not sure.